corehq/apps/data_interfaces/models.py

```diff
@@ -1214,6 +1324,101 @@ def to_dict(self):
         }
 
 
+class CaseDuplicateNew(models.Model):
+    id = models.BigAutoField(primary_key=True)
+    case_id = models.CharField(max_length=126, db_index=True)
+    action = models.ForeignKey("CaseDeduplicationActionDefinition", on_delete=models.CASCADE)
+    match_values = models.CharField(max_length=256)
+
+    class Meta:
+        db_table = "data_interfaces_caseduplicate_new"
+        unique_together = ('case_id', 'action')
+        indexes = [
+            models.Index(fields=['match_values', 'action_id'])
+        ]
+
+    def __str__(self):
+        return (
+            f"CaseDuplicateNew("
+            f"case_id={self.case_id}, action_id={self.action_id}, match_values={self.match_values})"
+        )
+
+    def delete(self, *args, check_for_orphans=True, **kwargs):
+        with transaction.atomic():
+            if check_for_orphans:
+                other_records = CaseDuplicateNew.objects.filter(
+                    action=self.action, match_values=self.match_values).exclude(case_id=self.case_id)[:2]
+
+                if other_records.count() == 1:
+                    # This will be orphaned when the current record is deleted, so delete it as well
+                    other_records[0].delete(check_for_orphans=False)
+
+            return super().delete(*args, **kwargs)
+
+    @classmethod
+    def create(cls, case, action, save=True):
+        match_values = cls.case_and_action_to_hash(case, action)
+        obj = cls(case_id=case.case_id, action=action, match_values=match_values)
+        if save:
+            obj.save()
+        return obj
+
+    @classmethod
+    def get_case_ids(cls, rule_id):
+        """Given a AutomaticUpdateRule id, return all case_ids that match
+        """
+        try:
+            rule = AutomaticUpdateRule.objects.get(
+                id=rule_id,
+                workflow=AutomaticUpdateRule.WORKFLOW_DEDUPLICATE,
+                deleted=False
+            )
+        except AutomaticUpdateRule.DoesNotExist:
+            return []
+        action_id = CaseDeduplicationActionDefinition.from_rule(rule).id
+        return list(cls.objects.filter(action_id=action_id).values_list('case_id', flat=True))
+
+    @classmethod
+    def remove_duplicates_for_case_ids(cls, case_ids):
+        entries = cls.objects.filter(case_id__in=case_ids)
+        for entry in entries:
+            # Individually delete models, rather than use a bulk query, so that the custom delete logic triggers
+            entry.delete()
```

Comment:
What is the largest expected number of items in `entries`? This seems pretty inefficient given that it will execute up to 3N queries. It may be possible to do this with as few as two queries. Probably only one with common table expressions.

Reply:
I never want to assume that the way the code is used will remain the same, but this method is currently only used when `CommCareUser.delete_user_data()` is called, and that iterates through all the user's case ids, 50 at a time.  While I can see a user having a large number of cases, I'd expect user deletion to occur rarely. I'd like to see what fewer queries would look like, but wouldn't that come at the expense of readability? Does this feel like premature optimization to you? O(3N) is still O(N)

Reply:
> wouldn't [fewer queries] come at the expense of readability?

Possibly. However, I think it could change O(N) to O(1), assuming all deletions for a given user could be done with one or two queries and the dominating performance factor is the number of database hits. Is that potentially a worthwhile optimization if this has a strong negative effect on deleting users with many cases?

Reply:
as I mentioned in the comment to the `_delete_with_orphan_check` suggestion, I'm concerned about the readability. This can be something we keep in mind should this prove to be a performance issue, but deleting users is such a rare activity, and when it does happen, it is not typically considered a performance-intensive one, that I'd prefer to keep the code as-is.

Reply:
I am concerned about scenarios like a project ending and requesting to delete all associated users (think USH). Depending on the number of users and cases involved, it seems like database performance could become a significant issue, and if that happened it would likely impact all active customers.